src/node/rpc/frontend.h

```diff
@@ -222,14 +224,32 @@ namespace ccf
             incoming_interface :
             interface_it.value().get<std::string>();
 
-          const auto primary_id = consensus->primary();
-          if (!primary_id.has_value())
+          std::optional<ccf::NodeId> target_node_id;
+          const auto nodes = InternalTablesAccess::get_trusted_nodes(tx);
+          {
+            const auto primary_id = consensus->primary();
+            if (seeking_primary)
+            {
+              target_node_id = primary_id;
+            }
+            else if (seeking_backup)
+            {
+              for (const auto& [id, _] : nodes)
+              {
+                if (id != primary_id)
+                {
+                  target_node_id = id;
```

Comment:
Just curious.

1. Do we redirect to the same node each time? The return value of `get_trusted_nodes` is a sorted (by id) map.
2. Why don't we exclude the current node from the search?

Reply:
1. Yes, we do currently direct to the same node each time. I think we should probably randomise this if we're trying to distribute load, and @achamayou suggested [here](https://github.com/microsoft/CCF/pull/6275#discussion_r1642724711) we should also prefer up-to-date nodes. I'll add that now.
2. We should only enter this path at all if the current node isn't a suitable target, though admittedly it's a little distant (we checked `is_primary` earlier, and we're checking if each ID matches the `primary()` ID here). I think this means we can't self-redirect currently, but does mean a small code change in future could produce self-redirections. Hopefully the fix suggested for part 1 also mitigates this. I'm not sure which error case is worse - producing a self-redirection, or deciding there is no redirect target. If a self-redirection happens because of a temporary flakiness (ie - we thought we couldn't handle this request initially, but by the time we're producing the redirect we look viable enough), then self-redirect seems useful - it probably waits out the temporary issue. But it could lead to an infinite redirect loop. I think while its feasible for a self-redirect to be valid (ie - I don't _think_ the state can change in this gap, but maybe it can, or will be able to in future), I'd err towards allowing self-redirects, and trust that a client recognises (and forbids) infinite redirect loops.